karapace/schema_registry.py

```diff
@@ -145,130 +143,131 @@ async def get_master(self, ignore_readiness: bool = False) -> Tuple[bool, Option
+            schemas = self.database.find_schemas(include_deleted=include_deleted, latest_only=latest_only)
+            return schemas
 
     def schemas_get(self, schema_id: int, *, fetch_max_id: bool = False) -> Optional[TypedSchema]:
-        with self.schema_reader.id_lock:
-            try:
-                schema = self.schema_reader.schemas.get(schema_id)
+        try:
+            schema = self.database.find_schema(schema_id=schema_id)
 
-                if schema and fetch_max_id:
-                    schema.max_id = self.schema_reader.global_schema_id
+            if schema and fetch_max_id:
+                schema.max_id = self.database.global_schema_id
 
-                return schema
-            except KeyError:
-                return None
+            return schema
+        except KeyError:
+            return None
 
     async def subject_delete_local(self, subject: str, permanent: bool) -> List[ResolvedVersion]:
         async with self.schema_lock:
-            subject_data_all = self.subject_get(subject, include_deleted=True)
-            subject_schemas_all = subject_data_all["schemas"]
+            schema_versions = self.subject_get(subject, include_deleted=True)
 
             # Subject can be permanently deleted if no schemas or all are soft deleted.
             can_permanent_delete = not bool(
-                [version for version, value in subject_schemas_all.items() if not value.get("deleted", False)]
+                [version_id for version_id, schema_version in schema_versions.items() if not schema_version.deleted]
             )
             if permanent and not can_permanent_delete:
                 raise SubjectNotSoftDeletedException()
 
             # Subject is soft deleted if all schemas in subject have deleted flag
-            already_soft_deleted = len(subject_schemas_all) == len(
-                [version for version, value in subject_schemas_all.items() if value.get("deleted", False)]
+            already_soft_deleted = len(schema_versions) == len(
+                [version_id for version_id, schema_version in schema_versions.items() if schema_version.deleted]
             )
-
             if not permanent and already_soft_deleted:
                 raise SubjectSoftDeletedException()
 
-            latest_schema_id = 0
+            latest_version_id = 0
             version_list = []
             if permanent:
-                version_list = list(subject_schemas_all)
-                latest_schema_id = version_list[-1]
-                for version, value in list(subject_schemas_all.items()):
-                    schema_id = value.get("id")
-                    LOG.info("Permanently deleting subject '%s' version %s (schema id=%s)", subject, version, schema_id)
+                version_list = list(schema_versions)
+                latest_version_id = version_list[-1]
+                for version_id, schema_version in list(schema_versions.items()):
```

Comment:
question (non-blocking):  I wonder why we have `list()` around items here?

Reply:
Not yet looking deeply, but I recall that the `schema_versions` is mutated from two threads. Creating the list here snapshots the items and concurrent modification error is avoided.

Reply:
Ah, makes sense 👍